subsys/usb/device_next/class/usbd_uvc.c

```diff
@@ -0,0 +1,2273 @@
+		uvc_flush_queue(dev);
+	} else {
+		LOG_WRN("Request on unknown endpoint 0x%02x", bi.udc.ep);
+	}
+
+	return 0;
+}
+
+/*
+ * Handling the start of USB transfers marked by 'v' below:
+ * v                                         v
+ * [h+data:][data:::][data:::]...[data:::]...[h+data:][data:::][data:::]...[data:::]...
+ * [vbuf:::::::::::::::::::::::::::::::::]...[vbuf:::::::::::::::::::::::::::::::::]...
+ * [frame::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::]...
+ */
+static struct net_buf *uvc_initiate_transfer(const struct device *dev, struct video_buffer *vbuf,
+					     size_t *next_line_offset, size_t *next_vbuf_offset)
+{
+	const struct uvc_config *cfg = dev->config;
+	struct uvc_data *data = dev->data;
+	struct video_format *fmt = &data->video_fmt;
+	size_t mps = uvc_get_bulk_mps(cfg->c_data);
+	struct net_buf *buf;
+
+	buf = net_buf_alloc_len(&uvc_buf_pool, mps, K_NO_WAIT);
+	if (buf == NULL) {
+		LOG_DBG("Cannot allocate first USB buffer for now");
+		return NULL;
+	}
+
+	if (fmt->pitch > 0) {
+		*next_line_offset += vbuf->bytesused / fmt->pitch;
+	}
+
+	LOG_INF("Start of transfer, bytes used %u, sending lines %u to %u out of %u",
+		vbuf->bytesused, vbuf->line_offset, *next_line_offset, fmt->height);
+
+	/* Copy the header into the buffer (first 2 fields only) */
+	net_buf_add_mem(buf, &data->payload_header, 2);
+	net_buf_add(buf, CONFIG_USBD_VIDEO_HEADER_SIZE - buf->len);
+
+	/* Copy the payload into the buffer */
+	*next_vbuf_offset = MIN(vbuf->bytesused, net_buf_tailroom(buf));
+	net_buf_add_mem(buf, vbuf->buffer, *next_vbuf_offset);
+
+	/* If this new USB transfer will complete this frame */
+	if (fmt->pitch == 0 || *next_line_offset >= fmt->height) {
+		LOG_DBG("Last USB transfer for this buffer");
+
+		/* Flag that this current transfer is the last */
+		((struct uvc_payload_header *)buf->data)->bmHeaderInfo |=
+			UVC_BMHEADERINFO_END_OF_FRAME;
+
+		/* Toggle the Frame ID of the next vbuf */
+		data->payload_header.bmHeaderInfo ^= UVC_BMHEADERINFO_FRAMEID;
+
+		*next_line_offset = 0;
+	}
+
+	return buf;
+}
+
+/*
+ * Handling the start of USB transfers marked by 'v' below:
+ *          v        v           v                    v        v           v
+ * [h+data:][data:::][data:::]...[data:::]...[h+data:][data:::][data:::]...[data:::]...
+ * [vbuf:::::::::::::::::::::::::::::::::]...[vbuf:::::::::::::::::::::::::::::::::]...
+ * [frame::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::]...
+ */
+static struct net_buf *uvc_continue_transfer(const struct device *dev, struct video_buffer *vbuf,
+					     size_t *next_line_offset, size_t *next_vbuf_offset)
+{
+	const struct uvc_config *cfg = dev->config;
+	struct uvc_data *data = dev->data;
+	size_t mps = uvc_get_bulk_mps(cfg->c_data);
+	struct net_buf *buf;
+
+	/* Directly pass the vbuf content with zero-copy */
+	buf = net_buf_alloc_with_data(&uvc_buf_pool, vbuf->buffer + data->vbuf_offset,
```

Comment:
Make sure that the buffer meets the requirements of the UDC controller/SoC, alignment and granularity (UDC_BUF_ALIGN/UDC_BUF_GRANULARITY). It might also make sense to have a separate pool for "zero-copy". If this is not yet possible, then you will need to bounce here.

Reply:
If you struggle to fix it, you can leave a comment about this limitation and address it later. One solution would be to submit compliant buffers or check the buffers and bounce them if necessary, but checks and memcopy would impact performance.